### ISSUE TITLE
No way to start an application with arguments

### ISSUE THREAD

**Issue by nathanwonnacott** [altranpt/Robot-Framework-Test-FX-Library#7]
As far as I can tell, there is no way to start a JavaFX application with command line arguments with this library. I think that this is a significant missing feature.

**Comment by nathanwonnacott** [comment]
Here's a potential bug I discovered that might be fixed at the same time that this new feature is implemented (assuming that you choose to implement it). If you don't implement this feature, then this should probably be a new issue as I believe this is a bug and should be fixed.

# The bug

According to [the docs](https://docs.oracle.com/javase/8/javafx/api/javafx/application/Application.html#getParameters--), `Application.getParameters()` should only return `null` if it is called from the application constructor. I have the following code:

```java
@Override
public void start(Stage primaryStage) throws Exception {
    String[] parameters = getParameters().getRaw().toArray(new String[0]);
    //More code
}
```

When calling the TestFXLibrary's `Start Application` keyword with this application, I get a null pointer exception because `getParameters()` returns `null`. It works just fine when I launch the app from Eclipse with no arguments (`getParameters().getRaw()` just returns an empty list). Is the TestFXLibrary launching the application with some sort of non-standard application life-cycle?

**Comment by jdagomes** [comment]
Hello @nathanwonnacott thank you for your contribution.

Regarding the bug you are describing, we are going to need a little bit more information, but what it seems is that you're trying to use the "Start Application" without compiling the your Java FX application. To test an application, you need to set the application jar files in the classpath. There is detailed information about this in [here](https://github.com/Altran-PT-GDC/Robot-Framework-Test-FX-Library/wiki/4.-Executing-Test-Scripts).

About the command line arguments to start a JavaFX application through the library, there is no point on doing that. The point of this library is to test JavaFX applications using Robot Framework, so you can use Robot Framework command line arguments, that are [here](http://robotframework.org/robotframework/latest/RobotFrameworkUserGuide.html#all-command-line-options).

If you have anything to add please feel free to do so. If this answered your post comment here so I don't close the issue.

Thank you for your contribution.

**Comment by jdagomes** [comment]
Hi @nathanwonnacott.

I will try to implement a new start application keyword with custom arguments. I'll keep you up to date in this topic as soon as I have news about this.

Thank you for your contribution

**Comment by nathanwonnacott** [comment]
Thanks, @jdagomes 

From your first comment, I don't think you fully understood my issue. From your second comment it looks like you got it.

Just to make sure I'm clear, this is what I want to be able to do the following (as an example):

# Java Code

```java
/**
 * SuperCoolFxApp
 * Created on Sep 17, 2018
 *
 * © Lockheed Martin Corporation, (2018). All rights reserved
 * Lockheed Martin Proprietary Information
 */

package com;

import java.io.File;
import java.util.List;

import javafx.application.Application;
import javafx.scene.Scene;
import javafx.scene.control.TextField;
import javafx.scene.layout.StackPane;
import javafx.stage.Stage;

public class SuperCoolFxApp extends Application {

	@Override
	public void start(Stage primaryStage) throws Exception {
		StackPane root = new StackPane();
		
		//Just display a text field with text describing where the
		//configuration for this app is coming from (based on a 
		//command line parameter passed to main(String args[]).
		List<String> args = this.getParameters().getRaw();
		
		String configurationFileDescription = "default config";
		
		//If there was an arg passed, use the first arg as a config file
		if(args.size() > 0) {
			configurationFileDescription = args.get(0);
			if(! (new File(configurationFileDescription).exists())) {
				configurationFileDescription = "Failed to load configuration";
			}
		}
		
		//TODO actually load the configuration data
		
		TextField txt = new TextField();
		txt.setId("ConfigurationTextBox");
		txt.setText(configurationFileDescription);
		root.getChildren().add(txt);
		
		primaryStage.setScene(new Scene(root, 200, 200));
		primaryStage.show();
	}

	public static void main(String args[]) {
		SuperCoolFxApp.launch(args);
	}
}

```

# Robot Code

```
*** Settings ***
Library    TestFXLibrary

*** Test Cases ***

Test Default Config
    Start Application    com.SuperCoolFxApp
    Select Main Window
    Text Field Text Should Be    ConfigurationTextBox    default config
    
Test Valid Config
    Start Application With Arguments    com.SuperCoolFxApp    aFileThatExists.xml
    Select Main Window
    Text Field Text Should Be    ConfigurationTextBox    aFileThatEXists.xml
    
Test Invalid Config
    Start Application With Arguments    com.SuperCoolFxApp    someFileThatDoesNotExist.xml
    Select Main Window
    Text Field Text Should Be    ConfigurationTextBox    Failed to load configuration
```

Right now, not only does the `Start Application With Arguments` keyword not exist, but in the `Test Default Config` method, a null pointer exception is thrown in the java code during the `getParameters().getRaw()` call, even though it should return an empty list.

I believe that the new keyword should look something like this:

```java
@RobotKeyword("Launches a JavaFX application with specified args")
@ArgumentNames({"mainClass", "args="})
public void startApplicationWithArguments(String mainClass, List<String> args) throws Exception {

	Class<?> clazz = Class.forName(mainClass);
	Method mainMethod = clazz.getMethod("main", String[].class);

	mainMethod.invoke(null, args.toArray());
}
```

**Comment by nathanwonnacott** [comment]
After looking at existing keywords, the ultimate implementation for this keyword will probably look something like this (in `com.altran.gdc.robotframework.testfxlibrary.keywords.misc):

```java
@RobotKeyword()
@ArgumentNames({"mainClass", "distinctiveName", "args="})
public void startApplicationWithArguments(String className, String distinctiveName, String... args){
        TestFxLibraryValidation.validateArguments(className, distinctiveName);

    final Stage[] stage = {null};
    try {
        int size = new FxRobot().listTargetWindows().size();
        if(size == 0) {
            FxToolkit.registerPrimaryStage();
            FxToolkit.setupApplication((Class<? extends Application>) Class.forName(className), args);
            FxToolkit.showStage();
            TestFXLibraryCache.getIstance().put(KEY + distinctiveName, FxToolkit.toolkitContext().getPrimaryStageFuture());
        } else {
            FxToolkit.registerStage(() -> {
                stage[0] = new Stage();
                return stage[0];
            });
            FxToolkit.setupApplication((Class<? extends Application>) Class.forName(className), args);
            FxToolkit.showStage();
            TestFXLibraryCache.getIstance().put(KEY + distinctiveName, stage[0]);
        }

    } catch (TimeoutException | ClassNotFoundException e) {
        throw new TestFxLibraryFatalException(e);
    }
}
```

However, when I try that out, the args still don't get passed to my application, so there must still be something wrong with this (or I'm doing something wrong in the software under test).

Also, this doesn't allow to make the `distinctiveName` parameter optional, so it's not ideal.

**Comment by nathanwonnacott** [comment]
This may require the use of a newer version of the TestFX library. Based on the `pom.xml` file, this library currently uses TestFX-4.0.6. In that version, my above implementation of the `StartApplicaitonWithArguments` keyword ends up calling the following method in `ToolkitServiceImpl`:

```java
    @Override
    public Future<Application> setupApplication(Supplier<Stage> stageSupplier,
                                                Class<? extends Application> applicationClass,
                                                String... applicationArgs) {
        return async(() -> {
            Application application = applicationService.create(() ->
                createApplication(applicationClass)
            ).get();
            applicationService.init(application).get();
            applicationService.start(application, stageSupplier.get()).get();
            return application;
        });
    }
```

As you can see, `applicationArgs` is passed in, but not actually used. In the current version of TestFX, the same method is:

```java
    @Override
    public Future<Application> setupApplication(Supplier<Stage> stageSupplier,
                                                Class<? extends Application> applicationClass,
                                                String... applicationArgs) {
        return async(() -> {
            Application application = asyncFx(() -> createApplication(applicationClass)).get();
            registerApplicationParameters(application, applicationArgs);
            applicationService.init(application).get();
            applicationService.start(application, stageSupplier.get()).get();
            return application;
        });
    }
```

I haven't followed the code through to see what `registerApplicationParameters(application, applicationArgs);` actually ends up doing, but it appears that the old version was ignoring the application args and the new version is at least attempting to use them properly.

**Comment by jdagomes** [comment]
Hi @nathanwonnacott.

It is working now with custom arguments. Checkout the "add_arguments_to_start_application" branch. Not going to merge it to master until it's all tested and updated TestFX to latest version, but you can checkout the branch and verify if it works for you.

What I have changed:

- TestFX version to 4.0.10-alpha (did not update to latest version because it is not working, will figure why it is failing and update to the latest version if possible);
- Add args to start application keyword and to FxToolkit.setupApplication (as your keyword suggestion);
- Add possibility to pass arguments to the  FxApplicationUnitTest application, changing the default text box text
- Add a test case to test the "Start Application" keyword with custom arguments, named "Test Start Application With Custom Arguments", check testFXLibraryMisc.txt

Thank you for your contribution, please check if it works in your application and give us feedback.

**Comment by nathanwonnacott** [comment]
@jdagomes 

I appreciate your incredibly quick turnaround time on this. I apologize for my not so incredibly quick turnaround on testing this...

Your fix seems to work great for me. One question though: does this now mean that the `distinctiveName` is now a required parameter if you want to specify arguments? If that's the case, then that's still fine. If I don't need a distinctiveName, I can just pass some bogus name.

**Comment by jdagomes** [comment]
Hi @nathanwonnacott,

The `distinctiveName` is not a required parameter, but you need to identify the custom arguments with `args=<your_args>` in Robot Framework test case, otherwise, because the `distinctiveName` is the second parameter of the `Start Application` keyword, it will recognize your arguments as the second parameter.

Example:
`     Start Application   testapp.FxApplicationUnitTest    args=myCustomArg`

As soon as this is fully tested and merged with the master branch, I will create a new release and update the documentation regarding this topic.

**Comment by codecube** [comment]
It is possible to put the distinctive_name as last parameter and pass
nothing (null)?

**Comment by codecube** [comment]
Another aproach would be to use a different method name with a specific
number of parameter (for ex. startApplicationWithDistinctiveName), that
would finally call the method with the named parameter.
If the parameter types are different, you could try use overlading,
although I don't know how it behaves in the case of Jython.

**Comment by nathanwonnacott** [comment]
Oops, I overlooked something in my testing, it appears that as it currently stands, this keyword will not work if you want to pass in more than one argument.

For example, if I run:

```
Start Application    testapp.FxApplicationUnitTest    someName    arg1    arg2
```

robot framework will say:

```
Keyword 'TestFXLibrary.Start Application' expected 1 to 3 arguments, got 4.
```

I believe that the code in Misc.java needs to be changed to:

```java
 @RobotKeyword
 @ArgumentNames({"className" , "distinctiveName=null", "*args=null"})
    public void startApplication(String className, String distinctiveName, String... args){
        //...
    }
```
(The only change is adding "*" before "args" in the argument names annotation).

However, this seems to mess up the named parameter capability, because now if I do:

```
@{args}=    Create List    arg1    arg2
Start Application    testapp.FxApplicationUnitTest    args=@{args}
```

Then on the java side, `distinctiveName` will be set to `args=[u'arg1', u'arg2']` and `args` will be an empty array.

I also tried with `args=${args}` instead of `args=@{args}` with the same result.

As far as I can tell, named parameters and varargs just don't get along. If there's no way to make named parameters and varargs work together, then I would say that it's best to just make `distinctiveName` and `args` both be required parameters.

**Comment by jdagomes** [comment]
I have updated the `add_arguments_to_start_application` branch, check the latest commit. In Misc.txt, there is a test named `Test Start Application With Custom Arguments` that tests starting a JavaFX application with multiple arguments.

Added `@ArgumentNames({"className" , "distinctiveName=null", "*args=null"})` to the keyword.
With `Start Application` keyword the way it is, to call multiple application arguments you need to give a `distinctiveName` parameter, like this:
`     Start Application   testapp.FxApplicationUnitTest    Application1    ArgumentNumber1    ArgumentNumber2    ArgumentNumber3`

@codecube Do you think it is better to have a different keyword for custom arguments, like `Start Application With Custom Arguments`, or having the `Start Application` keyword doing all this parsing?

**Comment by nathanwonnacott** [comment]
@jdagomes @codecube 
Is there any update on this?

**Comment by jdagomes** [comment]
Hi @nathanwonnacott, the new release with the arguments in Start Application is not yet  finished, I'll try to finish it this weekend and will notify you as soon as it is done.

**Comment by jdagomes** [comment]
Hi @nathanwonnacott ,

The new release 1.0.2 with custom arguments in Start Application keyword has been released [here](
https://github.com/Altran-PT-GDC/Robot-Framework-Test-FX-Library/releases/tag/1.0.2) and is now on the master branch.

You can set up a list with multiple arguments, check the example test case in `src\test\resources\robotframework\testFXLibraryMisc.txt` in `Test Start Application With Custom Arguments`

I'm closing this issue as it is fixed.

Thank you for your contribution!

**Comment by nathanwonnacott** [comment]
@jdagomes @codecube 
I've run into an issue with this keyword, but it isn't really your fault, so I'll leave it up to you whether you want to reopen this, create a new issue, or just leave things the way they are.

It turns out that the JavaLibCore library has a bug (in my opinion) that causes the Start Application keyword to work sometimes, but not other times (depending on the platform, and in some cases, luck). To make a long story short, with the current JavaLibCore library, on some platforms, it might call the startApplication(String) method even if you pass in arguments.

I've created an [issue](https://github.com/robotframework/JavalibCore/issues/21) in the JavaLibCore GitHub page. In my opinion, the *best* fix would be if the folks who maintain JavaLibCore would just fix it. However, it looks like it hasn't been updated in a couple of years and they don't seem super responsive to issues. So the other possible fix would be if you changed the name of the `Start Application` overload that accepts arguments to `Start Application With Args` or something like that.

It's up to you guys if you want to make the change or not. If you don't want to change anything, I've already built my own custom version of JavaLibCore that fixes the bug and I can just use that. However, if other users start using the `Start Application` keyword with arguments, they may run into the same issue.
